jscomp/bin/reactjs_jsx_ppx_v2.bspp.ml

```diff
@@ -312,7 +312,75 @@ let jsxMapper () =
   { default_mapper with structure; expr }
 #end
 
-#if defined BS_NO_COMPILER_PATCH then
+#if BS_COMPILER_IN_BROWSER then
+let rewrite code =
+  let mapper = jsxMapper () in
+  Location.input_name := "//toplevel//";
+  try 
+    let lexer = Lexing.from_string code in
+    let pstr = Parse.implementation lexer in
+    let pstr = mapper.structure mapper pstr in
+    Pprintast.structure Format.str_formatter pstr;
+    let ocaml_code = Format.flush_str_formatter () in
+    Printf.sprintf "{\"ocaml_code\": %S}" ocaml_code
+  with e ->
+    match Location.error_of_exn e with 
+    | Some error ->
+        let (file, line, startchar) =
+          Location.get_pos_info error.loc.loc_start in
+        let (file, endline, endchar) =
+          Location.get_pos_info error.loc.loc_end in
+        let result =
+          Printf.sprintf
+            "{ \
+              \"ppx_error_msg\": %S, \
+              \"row\": %d, \
+              \"column\": %d, \
+              \"endRow\": %d, \
+              \"endColumn\": %d, \
+              \"text\": %S, \
+              \"type\": \"error\" \
+            }"
+            (Printf.sprintf "Line %d, %d: %s" line startchar error.msg)
+            (line - 1) startchar (endline - 1) endchar error.msg in
+        Location.report_error Format.std_formatter error;
+        result
+    | None ->
+        Printf.sprintf "{\"js_error_msg\": %S}" (Printexc.to_string e)
+
+module Js = struct
```

Comment:
@bobzhang are you ok with this here? Seems copied over from jsoo

Reply:
Yes, it's based on https://github.com/BuckleScript/bucklescript/blob/master/jscomp/core/jsoo_main.ml#L49 (to provide a consistent API). Let me know if I have to modify it.